dart/lib/src/reference.dart

```diff
@@ -0,0 +1,416 @@
+    }
+    if (key is String && _valueType == ValueType.Map) {
+      var index = _keyIndex(key);
+      if (index != null) {
+        return _valueForIndex(index);
+      }
+    }
+    return null;
+  }
+
+  /// Get an iterable if the underlying flexBuffer value is a vector.
+  /// Otherwise throws an exception.
+  Iterable<Reference> get vectorIterable {
+    if(isVector == false) {
+      throw Exception('Value is not a vector. It is: ${_valueType}');
+    }
+    return _VectorIterator(this);
+  }
+
+  /// Get an iterable for keys if the underlying flexBuffer value is a map.
+  /// Otherwise throws an exception.
+  Iterable<String> get mapKeyIterable {
+    if(isMap == false) {
+      throw Exception('Value is not a map. It is: ${_valueType}');
+    }
+    return _MapKeyIterator(this);
+  }
+
+  /// Get an iterable for values if the underlying flexBuffer value is a map.
+  /// Otherwise throws an exception.
+  Iterable<Reference> get mapValueIterable {
+    if(isMap == false) {
+      throw Exception('Value is not a map. It is: ${_valueType}');
+    }
+    return _MapValueIterator(this);
+  }
+
+  /// Returns the length of the the underlying FlexBuffer value.
+  /// If the underlying value is [null] the length is 0.
+  /// If the underlying value is a number, or a bool, the length is 1.
+  /// If the underlying value is a vector, or map, the length reflects number of elements / element pairs.
+  /// If the values is a string or a blob, the length reflects a number of bytes the value occupies (strings are encoded in utf8 format).
+  int get length {
+    if (_length != null) {
+      return _length;
+    }
+    // needs to be checked before more generic isAVector
+    if(ValueTypeUtils.isFixedTypedVector(_valueType)) {
+      _length = ValueTypeUtils.fixedTypedVectorElementSize(_valueType);
+    } else if(_valueType == ValueType.Blob || ValueTypeUtils.isAVector(_valueType) || _valueType == ValueType.Map){
+      _length = _readInt(_indirect - _byteWidth, BitWidthUtil.fromByteWidth(_byteWidth));
+    } else if (_valueType == ValueType.Null) {
+      _length = 0;
+    } else if (_valueType == ValueType.String) {
+      var indirect = _indirect;
+      var size_byte_width = _byteWidth;
+      var size = _readInt(indirect - size_byte_width, BitWidthUtil.fromByteWidth(size_byte_width));
+      while (_buffer.getInt8(indirect + size) != 0) {
+        size_byte_width <<= 1;
+        size = _readInt(indirect - size_byte_width, BitWidthUtil.fromByteWidth(size_byte_width));
+      }
+      _length = size;
+    } else if (_valueType == ValueType.Key) {
+      var indirect = _indirect;
+      var size = 1;
+      while (_buffer.getInt8(indirect + size) != 0) {
+        size += 1;
+      }
+      _length = size;
+    } else {
+      _length = 1;
+    }
+    return _length;
+  }
+
+  /// Returns a minified JSON representation of the underlying FlexBuffer value.
+  /// Primary use should be debugging.
```

Comment:
Why? If someone wants to convert to JSON in production, what should they do instead?

Reply:
When someone __needs__ JSON in production than this is the way to go. I noted that primary goal should be debugging because this method will materialise the whole underlying object tree, which can be expensive. The `Reference` instance itself is just a pointer into the already loaded buffer and there for very light weight. So if people don't really need JSON, they should not call this method. But if you have to do this due to business requirements then you have to do it :).

Reply:
Perhaps this comment could read something like:

```dart
/// Returns a minified JSON representation of the underlying FlexBuffer value.
///
/// This method involves materializing the entire object tree, which may be
/// expensive. It is more efficient to work directly with the other methods on this
/// class to access only the needed data.
```
And leave it to the developer to determine what modalities they use it in?